src/boot/StoreHydratedGate.js

```diff
@@ -0,0 +1,57 @@
+/* @flow strict-local */
+import React, { useCallback } from 'react';
+import type { Node } from 'react';
+
+import { useEdgeTriggeredEffect } from '../reactUtils';
+import { useGlobalSelector, useDispatch } from '../react-redux';
+import { getIsHydrated, getHasAuth } from '../selectors';
+import FullScreenLoading from '../common/FullScreenLoading';
+import { registerAndStartPolling } from '../events/eventActions';
+import { sendOutbox } from '../outbox/outboxActions';
+import { initNotifications } from '../notification/notifTokens';
+
+type Props = $ReadOnly<{|
+  children: Node,
+|}>;
+
+/**
+ * Where we prevent everything from rendering while waiting for rehydration.
+ *
+ * Also where we dispatch post-rehydrate actions.
+ */
+export default function StoreHydratedGate(props: Props): Node {
+  const { children } = props;
+
+  const dispatch = useDispatch();
+
+  const isHydrated = useGlobalSelector(getIsHydrated);
+  const hasAuth = useGlobalSelector(getHasAuth);
+
+  const init = useCallback(async () => {
+    // Init right away if there's an active, logged-in account.
+    // NB `getInitialRouteInfo` depends intimately on this behavior.
+    if (hasAuth) {
+      await dispatch(registerAndStartPolling());
+
+      // TODO(#3881): Lots of issues with outbox sending
+      dispatch(sendOutbox());
+
+      dispatch(initNotifications());
+    }
+  }, [dispatch, hasAuth]);
+
+  useEdgeTriggeredEffect(
```

Comment:
In the final version, where this is in `StoreHydratedGate` and so `isHydrated` may start out false, I agree `useEdgeTriggeredEffect` is a good fit.

Continuing on from the comment above, though: I think the dependency on `hasAuth` should be taken as a genuine breach of the Rules of Hooks -- one that we're deliberately choosing here -- even though the ESLint plugin doesn't spot it. So there should be a comment explaining why that's OK. I think the point is basically that if `hasAuth` becomes true *later*, that's going to be either via ACCOUNT_SWITCH or LOGIN_SUCCESS, and each of those is accompanied by its own invocation of basically the same things as are in this `init` function.